src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -274,12 +339,167 @@ public void WriteRawValue(string rawValue)
+            this.isWritingConsecutiveRawValuesAtStartOfArray = false;
+        }
+
+        /// <summary>
+        /// Asynchronously manually writes an item separator ',' to the output, bypassing the Utf8JsonWriter.
+        /// </summary>
+        /// <returns>Task respresenting the asynchronous operation.</returns>
+        private async ValueTask WriteItemSeparatorAsync()
+        {
+            await this.FlushAsync().ConfigureAwait(false);
+            await this.writeStream.WriteAsync(itemSeparator).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Mark that the writer has transitioned to write an object.
+        /// </summary>
+        private void EnterObjectScope()
+        {
+            this.isWritingConsecutiveRawValuesAtStartOfArray = false;
+            this.isWritingFirstElementInArray = false;
+            this.bitStack.PushTrue();
+        }
+
+        /// <summary>
+        /// Mark that the writer has transitioned to write an array.
+        /// </summary>
+        private void EnterArrayScope()
+        {
+            this.isWritingConsecutiveRawValuesAtStartOfArray = false;
+            this.isWritingFirstElementInArray = true;
+            this.bitStack.PushFalse();
+        }
+
+        /// <summary>
+        /// Exit the current object or array scope.
+        /// </summary>
+        private void ExitScope()
+        {
+            this.isWritingFirstElementInArray = false;
+            this.shouldWriteSeparator = false;
+            this.bitStack.Pop();
+        }
+
+        /// <summary>
+        /// Checks whether the writer is currently in an array scope.
+        /// </summary>
+        /// <returns>Whether the writer is currently in an array scope.</returns>
+        private bool IsInArray()
+        {
+            if (!TryPeekScope(out bool isInObject))
+            {
+                return false;
+            }
+
+            return !isInObject;
+        }
+
+        /// <summary>
+        /// Tries to get the current scope.
+        /// </summary>
+        /// <param name="isInObject">
+        /// True if the writer is currently in an object scope, false if the writer is in an array scope.</param>
+        /// <returns>True if the current scope was successfully determined, false if it was not able to determined the scope
+        /// (i.e. we're neither in an array or object).
+        /// </returns>
+        private bool TryPeekScope(out bool isInObject)
+        {
+            isInObject = false;
+            if (this.bitStack.CurrentDepth == 0)
+            {
+                return false;
+            }
+
+            // BitStack doesn't implement a Peek()
+            // method, so we pop then push back.
+            isInObject = this.bitStack.Pop();
+            if (isInObject)
+            {
+                this.bitStack.PushTrue();
+            }
+            else
+            {
+                this.bitStack.PushFalse();
+            }

```

Comment:
is this bit here necessary? From the code just above it, we check the `CurrentDepth` if it is `0` we automatically return false which means the writer is in the `ArrayScope`  otherwise we return true for `ObjectScope`. In this code segment we are just popping and pushing which is not used in determining the current scope.

Reply:
The return value is not the value that was peeked, it is an indicator of whether or not we were able to peek. When the `CurrentDepth` is `0`, the stack is empty, and therefore we can't peek so we return false, otherwise we return true. The value that was peeked is placed in the `isInObject` `out` variable. The reason we pop then push the value back is because `BitStack` doesn't have a `Peek()` method, so we simulate one by popping the pushing back the value we popped.